addon/helpers/execute.js

```diff
@@ -0,0 +1,26 @@
+import { helper } from '@ember/component/helper';
+import isPromise from '../utils/is-promise';
+
+export function execute(actions = []) {
+  return function() {
+    let invoke = function(acc, curr) {
+      if (isPromise(acc)) {
+        return acc.then(curr);
+      }
+
+      return curr();
+    };
+
+    return actions.reduce((acc, curr, idx) => {
+      if (idx === 0) {
+        return curr();
+      }
+
+      return invoke(acc, curr);
+    }, undefined);
+  };
+}
+
+export default helper(execute);
+
+
```

Comment:
So this is a copy of `queue` without args?  What is the pain of ignoring the args in each `action`?

Reply:
Yes that's correct, a copy but without args, mainly to give the template ergonomics, Currently there's no way to simple chain events giving the template full control of partial application without leaks, queue and pipe 'leaks' into the chain of fns, for example using `on` with queue helper leaks the Dom event to the chain, using pipe leaks the return value, which might actually cause troubles if any fn in the chain returns void and any of the next function args expects maybe params  to be defined or something, you can see execute as the `call` helper but with a promise aware chain of fns

 To chain perhaps unrelated stuff like closing a dialog, creating a contact, displaying a notification and also triggering `@onClick` and logging to Google Analytics via a service action, all of these fns may receive args in certain way or order so leaking may have undesirable effects

```typescript
function something(str, isValid) {
  console.log(str)
  if(isValid){
     this.model.save()
  }
}
```

```handlebars
<button {{on "click" (queue this.openModal (fn this.something "hey there")}} >

Click me 

<button />

```

The event will be leaked, and so isValid would be true... I can try to give more examples if that doesn't make sense yet?
 
In our use case we have a serious amount of fns down on the render tree which need to hook say to the onBlur of an input and it wouldn't be posible if we had to worry about how they are actually physically placed, we see them as kind of a series of reactions to an event which of them is entirely responsible of partially apply the needed params

Reply:
Maybe call helper may receive a chain? Would that make more sense? 🤔🤔

Reply:
Yup, totally agree, what do you think about sequence?

- sequence

🤔🤔🤔

Reply:
Also, it sounds like the functions you are using might be called in multiple places?  Otherwise, you could always align the arguments with the arity of arguments if only used in one spot?

I'm also thinking about this from a typescript point of view.  You have a list of expected arguments.  However, this will pass nothing and could show an error.  Typically, in this case you would compose multiple functions together while perhaps sharing some logic.

What are your thoughts?

Reply:
> Also, it sounds like the functions you are using might be called in multiple places? Otherwise, you could always align the arguments with the arity of arguments if only used in one spot?

Yes, pretty much. I use them sparingly across the app/engines in different ways, from JS and helpers, so having full control on the template is a must, to avoid foot guns.

> I'm also thinking about this from a typescript point of view. You have a list of expected arguments. However, this will pass nothing and could show an error. Typically, in this case you would compose multiple functions together while perhaps sharing some logic.

Honestly I haven't (probably should) used a lot of functional programming or composition of fns, in `.js` land in Ember. I often use the templates for that, but I agree, you compose multiple functions together taking the arity, order and such into account when needed to share logic or something in common, in this particular case, the only thing these fns have in common is that they were triggered by something... sometimes an event, sometimes a callback, basically like observer pattern, pretty much, these are subscribers fns. in a Template only Component land, pipe, queue, (sequence|execute) come pretty handy

Reply:
Another good approach could be to just add a `noargs` helper...

```hbs
<button {{on "click" (queue fn1 fn2 (noargs (fn fn3 myWantedArg)))}}>
```

```ts
function noargs(fn) {
  return () => fn()
}
```